File: frontend/src/services/projectSkillService.js
import apiClient, {
  API_PROJECT_SKILLS,
  API_PROJECT_SKILLS_COUNT,
  API_PROJECT_SKILLS_BY_PROJECT,
  API_PROJECT_SKILLS_BY_SKILL,
  API_PROJECT_SKILLS_COUNT_BY_PROJECT,
  API_PROJECT_SKILLS_COUNT_BY_SKILL,
  API_PROJECT_SKILL_BY_ID,
} from './api';

class ProjectSkillService {
  // Create a new project-skill relationship
  async createProjectSkill(projectSkillData) {
    try {
      const response = await apiClient.post(API_PROJECT_SKILLS, projectSkillData);
      return response.data;
    } catch (error) {
      console.error('Error creating project-skill relationship:', error);
      throw error;
    }
  }

  // Get all project-skill relationships with pagination
  async getProjectSkills(skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS, {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships:', error);
      throw error;
    }
  }

  // Get project-skill relationship by ID
  async getProjectSkillById(projectSkillId) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILL_BY_ID(projectSkillId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationship:', error);
      throw error;
    }
  }

  // Update project-skill relationship
  async updateProjectSkill(projectSkillId, projectSkillData) {
    try {
      const response = await apiClient.put(API_PROJECT_SKILL_BY_ID(projectSkillId), projectSkillData);
      return response.data;
    } catch (error) {
      console.error('Error updating project-skill relationship:', error);
      throw error;
    }
  }

  // Delete project-skill relationship
  async deleteProjectSkill(projectSkillId) {
    try {
      const response = await apiClient.delete(API_PROJECT_SKILL_BY_ID(projectSkillId));
      return response.data;
    } catch (error) {
      console.error('Error deleting project-skill relationship:', error);
      throw error;
    }
  }

  // Get project-skill relationships count
  async getProjectSkillsCount() {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS_COUNT);
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships count:', error);
      throw error;
    }
  }

  // Get project-skill relationships by project
  async getProjectSkillsByProject(projectId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS_BY_PROJECT(projectId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships by project:', error);
      throw error;
    }
  }

  // Get project-skill relationships by skill
  async getProjectSkillsBySkill(skillId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS_BY_SKILL(skillId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships by skill:', error);
      throw error;
    }
  }

  // Get project-skill relationships count by project
  async getProjectSkillsCountByProject(projectId) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS_COUNT_BY_PROJECT(projectId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships count by project:', error);
      throw error;
    }
  }

  // Get project-skill relationships count by skill
  async getProjectSkillsCountBySkill(skillId) {
    try {
      const response = await apiClient.get(API_PROJECT_SKILLS_COUNT_BY_SKILL(skillId));
      return response.data;
    } catch (error) {
      console.error('Error fetching project-skill relationships count by skill:', error);
      throw error;
    }
  }
}

export default new ProjectSkillService(); 

File: backend/src/services/project_skill_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from src.model.project_skill import ProjectSkill, ProjectSkillCreate, ProjectSkillResponse
from typing import List, Optional
import logging

logger = logging.getLogger(__name__)

class ProjectSkillService:
    """Service class for handling project-skill relationship business logic."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_project_skill(self, project_skill_data: ProjectSkillCreate) -> ProjectSkillResponse:
        try:
            db_project_skill = ProjectSkill(**project_skill_data.model_dump())
            self.db.add(db_project_skill)
            self.db.commit()
            self.db.refresh(db_project_skill)
            logger.info(f"Created project-skill relationship with ID: {db_project_skill.id}")
            return ProjectSkillResponse.model_validate(db_project_skill)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to create project-skill relationship due to integrity error: {e}")
            raise ValueError("Project-skill relationship creation failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to create project-skill relationship: {e}")
            raise ValueError(f"Project-skill relationship creation failed: {str(e)}")
    
    def get_project_skill_by_id(self, project_skill_id: int) -> Optional[ProjectSkillResponse]:
        project_skill = self.db.query(ProjectSkill).filter(ProjectSkill.id == project_skill_id).first()
        if project_skill:
            return ProjectSkillResponse.model_validate(project_skill)
        return None
    
    def get_project_skills_by_project(self, project_id: int, skip: int = 0, limit: int = 10) -> List[ProjectSkillResponse]:
        project_skills = self.db.query(ProjectSkill).filter(
            ProjectSkill.project_id == project_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_skills)} project-skill relationships for project {project_id}")
        return [ProjectSkillResponse.model_validate(ps) for ps in project_skills]
    
    def get_project_skills_by_skill(self, skill_id: int, skip: int = 0, limit: int = 10) -> List[ProjectSkillResponse]:
        project_skills = self.db.query(ProjectSkill).filter(
            ProjectSkill.skill_id == skill_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_skills)} project-skill relationships for skill {skill_id}")
        return [ProjectSkillResponse.model_validate(ps) for ps in project_skills]
    
    def list_project_skills(self, skip: int = 0, limit: int = 10) -> List[ProjectSkillResponse]:
        project_skills = self.db.query(ProjectSkill).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(project_skills)} project-skill relationships with skip={skip}, limit={limit}")
        return [ProjectSkillResponse.model_validate(ps) for ps in project_skills]
    
    def update_project_skill(self, project_skill_id: int, project_skill_data: ProjectSkillCreate) -> Optional[ProjectSkillResponse]:
        db_project_skill = self.db.query(ProjectSkill).filter(ProjectSkill.id == project_skill_id).first()
        if not db_project_skill:
            return None
        try:
            for key, value in project_skill_data.model_dump().items():
                setattr(db_project_skill, key, value)
            self.db.commit()
            self.db.refresh(db_project_skill)
            logger.info(f"Updated project-skill relationship with ID: {db_project_skill.id}")
            return ProjectSkillResponse.model_validate(db_project_skill)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to update project-skill relationship {project_skill_id} due to integrity error: {e}")
            raise ValueError("Update failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to update project-skill relationship {project_skill_id}: {e}")
            raise ValueError(f"Update failed: {str(e)}")
    
    def delete_project_skill(self, project_skill_id: int) -> bool:
        db_project_skill = self.db.query(ProjectSkill).filter(ProjectSkill.id == project_skill_id).first()
        if not db_project_skill:
            return False
        try:
            self.db.delete(db_project_skill)
            self.db.commit()
            logger.info(f"Deleted project-skill relationship with ID: {project_skill_id}")
            return True
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to delete project-skill relationship {project_skill_id}: {e}")
            raise ValueError(f"Deletion failed: {str(e)}")
    
    def get_project_skills_count(self) -> int:
        return self.db.query(ProjectSkill).count()
    
    def get_project_skills_count_by_project(self, project_id: int) -> int:
        return self.db.query(ProjectSkill).filter(ProjectSkill.project_id == project_id).count()
    
    def get_project_skills_count_by_skill(self, skill_id: int) -> int:
        return self.db.query(ProjectSkill).filter(ProjectSkill.skill_id == skill_id).count() 